### ISSUE TITLE
Ignored test in AsyncFunSuite breaks reporting for subsequent tests

### ISSUE THREAD

**Issue by sangamon** [sbt/sbt#5245]
## steps

sbt version: 1.3.2 (locally installed version), 1.2.8/1.3.3 (build.properties declarations), probably others

Scala version: 2.12.8/2.13.1

ScalaTest version: 3.0.8

1. create a new project with ScalaTest dependency
1. `build.sbt`: `parallelExecution in Test := false, fork in Test := true`
1. create two test classes extending `AsyncFunSuite`
1. in each test class, add three test cases
   - `test(...)` successful
   - `test(...)` failing
   - `ignore(...)`
   
   (see notes for details)
1. run test suite for the full package/src folder in IntelliJ IDEA (or other IDE)
1. run "sbt test"

## problem

IntelliJ IDEA will correctly report "Tests failed: 2, passed: 2, ignored: 2 of 6 tests".

`sbt test` output:
```
[info] IgnoreBugTestA:
[info] - a-succ
[info] - a-fail *** FAILED ***
[info]   failed (IgnoreBugTestA.scala:14)
[info] - a-ign !!! IGNORED !!!
[info] Run completed in 1 second, 567 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 0, aborted 0
[info] Tests: succeeded 1, failed 1, canceled 0, ignored 1, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	de.sangamon.ignorebug.IgnoreBugTestB
[error] 	de.sangamon.ignorebug.IgnoreBugTestA
[error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
```

## expectation

Proper reporting of tests from second class (`IgnoreBugTestB` in the above example).

In the real-world scenario that triggered this report, with no failures occurring, the test class containing the ignored case happens to be the first one executed. Instead of the expected summary
```
[info] Total number of tests run: 47
[info] Suites: completed 12, aborted 0
[info] Tests: succeeded 47, failed 0, canceled 0, ignored 1, pending 0
```
we get
```
[info] Total number of tests run: 11
[info] Suites: completed 0, aborted 0
[info] Tests: succeeded 11, failed 0, canceled 0, ignored 1, pending 0
```

## notes

I'm not sure whether the problem is with ScalaTest or with sbt, but since it only occurs from sbt, I'm reporting it here.

- if the ignored test case is removed, everything works as expected
- the behavior could not be observed using a plain `FunSuite`
- with `fork in Test := false`, the summary output is correct, single test case still aren't reported
- `logBuffered in Test := false` doesn't seem to make a difference

Full test class:
```
class IgnoreBugTestA extends AsyncFunSuite {

  test("a-succ") {
    Future.successful(succeed)
  }

  test("a-fail") {
    Future.successful(fail("failed"))
  }

  ignore("a-ign") {
    ???
  }

}
```
(`IgnoreBugTestB` accordingly.)

**Comment by sangamon** [comment]
[ignorebug.tar.gz](https://github.com/sbt/sbt/files/3887557/ignorebug.tar.gz)

**Comment by sangamon** [comment]
Attached complete sample project.

**Comment by eed3si9n** [comment]
@sangamon Thanks for the detailed report!

**Comment by eed3si9n** [comment]
To potential contributor: Relevant code might be around
- https://github.com/sbt/sbt/blob/develop/main-actions/src/main/scala/sbt/ForkTests.scala
- https://github.com/sbt/sbt/blob/develop/testing/src/main/scala/sbt/TestFramework.scala
- https://github.com/sbt/sbt/blob/develop/testing/src/main/scala/sbt/TestReportListener.scala
- https://github.com/scalatest/scalatest/blob/release-3.0.8/scalatest.js/src/main/scala/org/scalatest/tools/MasterRunner.scala
- https://github.com/scalatest/scalatest/blob/release-3.0.8/scalatest.js/src/main/scala/org/scalatest/tools/TaskRunner.scala

**Comment by johnduffell** [comment]
We are hitting the same issue, also affects sbt 1.3.12 and only affect AsyncFlatSpec, not the standard FlatSpec.  It doesn't happen with the scalatest runner, only the SBT one, so I think it's an sbt issue too.
It also affects excluded tests, e.g. via tags, as well as ignored tests.

The interesting thing I have discovered over and above your report is that once a test is ignored, it doesn't show the "suite completed" event when run in scalatest's "unformatted" mode.

Running correctly via the scalatest runner with `-eU` option:
```
Suite Starting - AnotherTest
Test Starting - AnotherTest: should NOT be ignored
Test Succeeded - AnotherTest: should NOT be ignored
Suite Completed - AnotherTest
Suite Starting - IgnoreTest
Test Ignored - IgnoreTest: should be ignored
Test Starting - IgnoreTest: should NOT be ignored
Test Succeeded - IgnoreTest: should NOT be ignored
Suite Completed - IgnoreTest
```
running via sbt with 
`testOptions in Test += Tests.Argument(TestFrameworks.ScalaTest, "-eU")`

will show either 
```
Suite Starting - AnotherTest
Test Starting - AnotherTest: should NOT be ignored
Test Succeeded - AnotherTest: should NOT be ignored
Suite Completed - AnotherTest
```
or
```
Suite Starting - AnotherTest
Test Starting - AnotherTest: should NOT be ignored
Test Succeeded - AnotherTest: should NOT be ignored
Suite Completed - AnotherTest
Suite Starting - IgnoreTest
Test Ignored - IgnoreTest: should be ignored
Test Starting - IgnoreTest: should NOT be ignored
Test Succeeded - IgnoreTest: should NOT be ignored
```
and never `Suite Completed - IgnoreTest`

So I think this is the root of the problem, sbt runner is not picking up the completion of the suite.

Simplest SBT project to reproduce is here:
[scalatest-async-bug.zip](https://github.com/sbt/sbt/files/4794460/scalatest-async-bug.zip)

**Comment by johnduffell** [comment]
I think the issue is on this line, ListBuffer `+=` doesn't look to be thread safe.
https://github.com/sbt/sbt/blob/9f9bb9daa275458f3e55f87c432c599947a71b69/testing/src/main/scala/sbt/TestFramework.scala#L123
Looks like a very similar issue has been fixed already
https://github.com/sbt/sbt/pull/4030
and a different fix here (looks like the contract on the eventhandler is not clear)
https://github.com/scala/scala-partest/pull/78
I noticed the problem doesn't happen when you stop in the debugger.

**Comment by johnduffell** [comment]
I think there are at least two bugs, one in scalatest that some events do not appear which I found a fix for:
https://github.com/scalatest/scalatest/pull/1841
This makes the test reporter show the right events, but there is still an issue with incomplete/non deterministic output when I remove the `-eU` flag, which may be an SBT or scalatest issue.

**Comment by Eruis2579** [comment]
@eed3si9n , Can I try this issue?

**Comment by eed3si9n** [comment]
@Eruis2579 Yes, please. Please check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) and:
1. Make sure you can reproduce the problem and the fix manually
2. Include tests for the change
3. Follow our AI assisted contributions guideline (please disclose AI usages in the PR description, if any)
4. Multiple contributors might work on this issue in parallel
5. Generally we recommend getting CI to pass on a forked repo before sending a pull request (pull requests that do not pass tests will be closed after a few days)
